pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2020 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     pkg_mbedtls
+ *
+ * @{
+ * @file
+ * @brief
+ *
+ * @author      Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
+ *
+ * @}
+ */
+#include <assert.h>
+
+#include "mbedtls/entropy.h"
+#include "mbedtls/entropy_poll.h"
+
+#include "entropy_sources_mbedtls_riot.h"
+
+#include "kernel_defines.h"
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_HWRNG)
+#include "periph/hwrng.h"
+#endif
+
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_ADC)
+#include "entropy_source/adc_noise.h"
+#endif
+
+/**
+ * @brief   Number of available entropy sources.
+ */
+#define NUM_ENTROPY_FUNCS ARRAY_SIZE(entropy_funcs)
+
+/**
+ * @brief   Configuration of available entropy sources.
+ *
+ * @note    First source must be the best source available. It is
+ *          expected to be strong.
+ * @note    An entropy source is configured if the platform provides
+            the underlying feature and the PSEUDOMODULE (which is
+            included by default) is not manually disabled.
+ * @{
+ */
+const entropy_source_mbedtls_riot_t entropy_funcs[] = {
+#ifdef MODULE_MBEDTLS_ENTROPY_SOURCE_HWRNG
+{ .func = riot_hwrng_poll, .strong = MBEDTLS_ENTROPY_SOURCE_STRONG },
+#endif
+#ifdef MODULE_MBEDTLS_ENTROPY_SOURCE_ADC
+{ .func = riot_adc_poll, .strong = MBEDTLS_ENTROPY_SOURCE_WEAK },
+#endif
+    /* Additional sources need to be added here */
+};
+/** @} */
+
+/* Connect first entropy source because this function
+ * is used for internal tests
+ */
+int mbedtls_hardware_poll(void *data, unsigned char *output, size_t len,
+                            size_t *olen)
+{
+    assert(output != NULL && olen!= NULL);
+
+    return entropy_funcs[0].func(data, output, len, olen);
+}
+
+/* Add more entropy sources. Start from index=1 because index=0 is used in
+ * mbedtls_hardware_poll already
+ */
+int riot_add_src_avail(mbedtls_entropy_context *ctx)
+{
+    int ret = 0;
+
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_ADC)
+    entropy_source_adc_init();
+#endif
```

Comment:
It depends on the source, but seems likely that most sources have some kind of an init function. Technically, we could add  .init function to `entropy_source_mbedtls_riot_t` but I would like to delay that, since other sources (e.g. PUF) might have very different requirements from ADC sampling or HWRNG readout.